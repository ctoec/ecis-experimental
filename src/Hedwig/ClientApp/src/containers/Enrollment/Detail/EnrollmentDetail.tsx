import React, { useContext } from 'react';
import ChildInfo from '../_sections/ChildInfo';
import FamilyInfo from '../_sections/FamilyInfo';
import FamilyIncome from '../_sections/FamilyIncome';
import EnrollmentFunding from '../_sections/EnrollmentFunding';
import { Link } from 'react-router-dom';
import { nameFormatter, enrollmentDetailMetadataFormatter } from '../../../utils/stringFormatters';
import {
	ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest,
	Enrollment,
} from '../../../generated';
import UserContext from '../../../contexts/User/UserContext';
import useApi from '../../../hooks/useApi';
import { validatePermissions, getIdForUser } from '../../../utils/models';
import { InlineIcon, Button } from '../../../components';
import CommonContainer from '../../CommonContainer';
import { SectionProps } from '../enrollmentTypes';

type EnrollmentDetailParams = {
	match: {
		params: {
			siteId: number;
			enrollmentId?: number;
		};
	};
};

const sections = [ChildInfo, FamilyInfo, FamilyIncome, EnrollmentFunding];

/**
 * React component for displaying enrollment summary information.
 * Hands off to section summary component.
 *
 * @param props Props with location
 */
export default function EnrollmentDetail({
	match: {
		params: { siteId, enrollmentId },
	},
}: EnrollmentDetailParams) {
	const { user } = useContext(UserContext);

	// Get enrollment by id
	const params: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
		id: enrollmentId ? enrollmentId : 0,
		orgId: getIdForUser(user, 'org'),
		siteId: validatePermissions(user, 'site', siteId) ? siteId : 0,
		include: ['child', 'family', 'determinations', 'fundings', 'sites'],
	};
	const [loading, error, enrollment, mutate] = useApi<Enrollment>(
		api => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		[enrollmentId, user],
		{
			skip: !enrollmentId,
		}
	);

	if (loading || error || !enrollment) {
		return <div className="EnrollmentDetail"></div>;
	}

	const child = enrollment.child;

	return (
		<CommonContainer
			directionalLinkProps={{
				direction: 'left',
				to: '/roster',
				text: 'Back to roster',
			}}
		>
			<div className="grid-container">
				<div className="grid-row flex-first-baseline flex-space-between">
					<div>
						<h1>{nameFormatter(child)}</h1>
						<span className="usa-hint text-italic">{enrollmentDetailMetadataFormatter(enrollment)}</span>
					</div>
					<Button text="Withdraw" href={`/roster/sites/${siteId}/enrollments/${enrollment.id}/withdraw`} className="margin-right-0" />
				</div>
				{sections.map(section => {
					var props: SectionProps = { siteId, enrollment, mutate };
					const familyIncomeForFosterChild = section.key === 'family-income' && child.foster;
					return (
						<section key={section.key} className="oec-enrollment-details-section">
							<div className="oec-enrollment-details-section__content">
								<h2>{section.name}</h2>
								<section.Summary {...props} />
							</div>
							<div className="oec-enrollment-details-section__actions">
								{section.status(props) === 'incomplete' && (
									<span>
										<InlineIcon icon="incomplete" /> Missing information
									</span>
								)}
								<Link
									to={`/roster/sites/${siteId}/enrollments/${enrollment.id}/edit/${section.key}`}
									className={familyIncomeForFosterChild ? 'display-none important' : ''}
								>
									Edit<span className="usa-sr-only"> {section.name.toLowerCase()}</span>
								</Link>
							</div>
						</section>
					);
				})}
			</div>
		</CommonContainer>
	);
}
