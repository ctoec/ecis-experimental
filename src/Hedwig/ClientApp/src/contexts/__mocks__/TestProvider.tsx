import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';

import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, ReportingPeriod } from '../../generated';
import { cdcReportingPeriods as _cdcReportingPeriods, user as _user } from '../../tests/data';

type TestProviderProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const TestProvider: React.FC<TestProviderProps> = ({
	children,
	user = _user,
	cdcReportingPeriods = _cdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<UserContext.Provider value={{ user }}>
			<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
				<Router history={history}>{children}</Router>
			</ReportingPeriodContext.Provider>
		</UserContext.Provider>
	);
};

export default TestProvider;
