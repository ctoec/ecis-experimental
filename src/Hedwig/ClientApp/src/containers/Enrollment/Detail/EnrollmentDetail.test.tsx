// Variables used in jest mockes -- must start with `mock`
import { mockAllFakeEnrollments, mockSite } from '../../../tests/data';
import mockUseApi, {
	mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet,
	mockApiOrganizationsOrgIdSitesIdGet,
} from '../../../hooks/useApi/__mocks__/useApi';

// Jest mocks must occur before later imports
jest.mock('../../../hooks/useApi', () =>
	mockUseApi({
		apiOrganizationsOrgIdSitesIdGet: mockApiOrganizationsOrgIdSitesIdGet(mockSite),
		apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet: mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(
			mockAllFakeEnrollments
		),
	})
);

import React from 'react';
import { render } from '@testing-library/react';
import EnrollmentDetail from './EnrollmentDetail';
import TestProvider from '../../../contexts/__mocks__/TestProvider';
import { mockCompleteEnrollment, mockEnrollmentMissingBirthCertId } from '../../../tests/data';
import { accessibilityTestHelper } from '../../../tests/helpers';

afterAll(() => {
	jest.resetModules();
});

describe('EnrollmentDetail', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(
			<TestProvider>
				<EnrollmentDetail
					match={{
						params: {
							siteId: mockCompleteEnrollment.siteId,
							enrollmentId: mockCompleteEnrollment.id,
						},
					}}
				/>
			</TestProvider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('shows incomplete indications when incomplete information is given', () => {
		const { getAllByText } = render(
			<TestProvider>
				<EnrollmentDetail
					match={{
						params: {
							siteId: mockEnrollmentMissingBirthCertId.siteId,
							enrollmentId: mockEnrollmentMissingBirthCertId.id,
						},
					}}
				/>
			</TestProvider>
		);

		const incompleteIcons = getAllByText('(incomplete)');
		expect(incompleteIcons.length).toBe(1);
	});

	accessibilityTestHelper(
		<TestProvider>
			<EnrollmentDetail
				match={{
					params: {
						siteId: mockEnrollmentMissingBirthCertId.siteId,
						enrollmentId: mockEnrollmentMissingBirthCertId.id,
					},
				}}
			/>
		</TestProvider>
	);
});
