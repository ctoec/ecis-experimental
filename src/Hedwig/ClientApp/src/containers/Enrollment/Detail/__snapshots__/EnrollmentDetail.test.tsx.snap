// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrollmentDetail matches snapshot 1`] = `
<div>
  <div
    class="grid-container"
  >
    <a
      class="oec-directional-link"
      href="/roster"
    >
      <svg
        class="oec-directional-link__arrow oec-directional-link__arrow--direction-left oec-directional-link__arrow--side-left"
      >
        arrowRight.svg
      </svg>
      Back to roster
    </a>
  </div>
  <div
    class="grid-container"
  >
    <div
      class="grid-row flex-first-baseline flex-space-between"
    >
      <div>
        <h1>
          Lily Luna Potter
        </h1>
        <span
          class="usa-hint text-italic"
        >
          Last updated on January 1st, 2020 by T. User
        </span>
      </div>
      <a
        class="usa-button margin-right-0"
        href="/roster/sites/1/enrollments/1/withdraw"
      >
        Withdraw
      </a>
    </div>
    <section
      class="oec-enrollment-details-section"
    >
      <div
        class="oec-enrollment-details-section__content"
      >
        <h2>
          Child information
        </h2>
        <div
          class="ChildInfoSummary"
        >
          <p>
            Name: 
            Lily Luna Potter
          </p>
          <p>
            Birthdate: 
            12/12/2016
          </p>
          <p>
            Birth certificate: 
            Yes
          </p>
          <p>
            Race: 
            Native Hawaiian Or Pacific Islander
          </p>
          <p>
            Ethnicity: 
            Hispanic/Latinx
          </p>
          <p>
            Gender: 
            Female
          </p>
        </div>
      </div>
      <div
        class="oec-enrollment-details-section__actions"
      >
        <a
          class=""
          href="/roster/sites/1/enrollments/1/edit/child-information"
        >
          Edit
          <span
            class="usa-sr-only"
          >
             
            child information
          </span>
        </a>
      </div>
    </section>
    <section
      class="oec-enrollment-details-section"
    >
      <div
        class="oec-enrollment-details-section__content"
      >
        <h2>
          Family information
        </h2>
        <div
          class="FamilyInfoSummary"
        >
          <p>
            Address: 
            4 Privet Drive, Hogsmeade, CT 77777
             
            
          </p>
        </div>
      </div>
      <div
        class="oec-enrollment-details-section__actions"
      >
        <a
          class=""
          href="/roster/sites/1/enrollments/1/edit/family-information"
        >
          Edit
          <span
            class="usa-sr-only"
          >
             
            family information
          </span>
        </a>
      </div>
    </section>
    <section
      class="oec-enrollment-details-section"
    >
      <div
        class="oec-enrollment-details-section__content"
      >
        <h2>
          Family income determination
        </h2>
        <div
          class="FamilyIncomeSummary"
        >
          <p>
            Income determination not disclosed.
          </p>
        </div>
      </div>
      <div
        class="oec-enrollment-details-section__actions"
      >
        <a
          class=""
          href="/roster/sites/1/enrollments/1/edit/family-income"
        >
          Edit
          <span
            class="usa-sr-only"
          >
             
            family income determination
          </span>
        </a>
      </div>
    </section>
    <section
      class="oec-enrollment-details-section"
    >
      <div
        class="oec-enrollment-details-section__content"
      >
        <h2>
          Enrollment and funding
        </h2>
        <div
          class="EnrollmentFundingSummary"
        >
          <p>
            Site: 
             
          </p>
          <p>
            Age group:
             
            Preschool
          </p>
          <p>
             
            Enrollment date:
             
            02/03/2018
             
          </p>
          <p>
            Funding:
             
            CDC - full time
          </p>
          <p>
            First reporting period:
             
            March 2019
          </p>
        </div>
      </div>
      <div
        class="oec-enrollment-details-section__actions"
      >
        <a
          class=""
          href="/roster/sites/1/enrollments/1/edit/enrollment-funding"
        >
          Edit
          <span
            class="usa-sr-only"
          >
             
            enrollment and funding
          </span>
        </a>
      </div>
    </section>
  </div>
</div>
`;
